packages/nuxt/src/app/composables/preview.ts

```diff
@@ -0,0 +1,65 @@
+import { reactive, readonly, toRef } from 'vue'
+
+import { refreshNuxtData, useRoute, useRouter, useState } from '#app'
+
+export interface PreviewOptions<Controls extends boolean = false> {
+  controls?: Controls
+  tokenQueryName?: string
+}
+
+interface PreviewState {
+  enabled: boolean
+  token: string | null
+  addedAfterNavigationCallback: boolean
+}
+
+const previewQueryName = 'preview'
+
+export function usePreviewMode<Controls extends boolean = false> (options?: PreviewOptions<Controls>) {
+  options = {
+    controls: false as Controls,
+    tokenQueryName: 'token',
+    ...(options || {})
+  }
+
+  const router = useRouter()
+  const route = useRoute()
+
+  const previewState = useState('_preview-composable', () => reactive<PreviewState>({
+    enabled: false,
+    token: null,
```

Comment:
A few ideas:
* instead of defaulting to just `token` which I agree is the most common scenario, what do you think about setting arbitrary state?
* What about an option that ignores query param and allowing user to manually engage/disengage query mode instead (for more complex situations? 
* when `enabled` is manually set to false, we probably want to remove the refreshNuxtData router hook - maybe a computed ref with manual set/get values could be useful here?

Reply:
I have a little question, is there a point of something like `shouldEnable` if we give user ability to toggle `on` and `off` preview mode ? I think of flag `shouldAutoDetect` and if it is `true`, then detect `preview` in query (or we could say that `shouldAutoDetect` can be `boolean` or `string`, if it is a string we will match for that string instead), if `false`, let user manually do some detection (in for example `onMounted` hook) and set `enabled` to true ?